### PR TITLE
Sinatra App Project

### PR DESCRIPTION
**Pull request by andersann**
Creating first app on sinatra

### CODE REVIEW

**Comment by stahor** [comment]
Ну для начала выложи ссылку на прилагу и краткое описание в, собственно, описание.

**Comment by andersann** [comment]
@stahor added some description in README file

File: app.rb [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    if !user_exist?(user.email) && user.save
+      session[:authenticated] = true
+      redirect '/users'
+    else
+      'Error'
+    end
+  end
+
+  def user_exist?(email)
+    User.find_by(email: email)
```
Comment (stahor) [inline app.rb]: use `exists?`

File: app.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+get '/' do
+  erb :index
+end
+
+get '/login' do
+  erb :login
+end
+
+post '/submit_login' do
+  if User.find_by(params[:user])
```
Comment (stahor) [inline app.rb]: WTF?! No password verification? Use warden gem or one of it's sinatra adapters for authentication logic.

File: app.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  set :inline_templates, true
+end
+
+helpers do
+  def save_user(user)
+    if !user_exist?(user.email) && user.save
+      session[:authenticated] = true
+      redirect '/users'
+    else
+      'Error'
```
Comment (stahor) [inline app.rb]: use `error` method

File: app.rb [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  if User.find_by(email: email)
+    session[:authenticated] = true
+    redirect '/users'
+  else
+    @user = User.new(email: email, name: name)
+    save_user(@user)
+  end
+end
+
+post '/registrate' do
```
Comment (stahor) [inline app.rb]: register

File: app.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            }
+          }
+        )
+      )
+  end
+
+  email = auth_result['private_emails'][0]['email']
+  name = auth_result['login']
+  if User.find_by(email: email)
+    session[:authenticated] = true
```
Comment (stahor) [inline app.rb]: I've seen this before. DRY

Comment (stahor) [inline app.rb]: You probably want to have `@current_user` available in each action of this app. How can you achieve this?

File: app.rb [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    'LogIn Error. User not found'
+  end
+end
+
+get '/registration' do
+  erb :registration
+end
+
+get '/callback' do
+  session_code = request.env['rack.request.query_hash']['code']
```
Comment (stahor) [inline app.rb]: Huge terrible method. Split it.

**Comment by stahor** [comment]
* Add passwords and confirmation
* Add bcrypt
* Add admin role

**Comment by andersann** [comment]
@stahor , I don't ignore, I'm correcting my mistakes little by little, basically we are working on a project now

**Comment by andersann** [comment]
@stahor new commit added.
On the next step I want to add user profile.

File: Gemfile [hunk 7]
```diff
@@ -0,0 +1,13 @@
+source "https://rubygems.org"
+
+gem 'sinatra'
```
Comment (andersann) [inline Gemfile]: @stahor this means adding versions to each gem?

Comment (stahor) [inline Gemfile]: yup

File: app.rsa.example [hunk 8]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (stahor) [inline app.rsa.example]: I hope this is not real key?

File: app.rb [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    rescue JWT::DecodeError
+      return false
+    end
+  end
+
+  def create_session
+    headers = { exp: Time.now.to_i + 200 }
+    token = JWT.encode({ user_id: @user.id }, settings.signing_key, 'RS256', headers)
+    session['access_token'] = token
+    puts session['access_token']
```
Comment (stahor) [inline app.rb]: puts?

File: app.rb [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+require 'jwt'
+
+CLIENT_ID = ENV['GH_BASIC_CLIENT_ID']
+CLIENT_SECRET = ENV['GH_BASIC_SECRET_ID']
+SESSION_SECRET = ENV['SESSION_SECRET']
+
+signing_key_path = File.expand_path('../app.rsa', __FILE__)
+verify_key_path = File.expand_path('../app.rsa.pub', __FILE__)
+
+signing_key = ''
```
Comment (stahor) [inline app.rb]: no need for them

File: app.rb [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+require 'json'
+require 'dotenv/load'
+require 'sinatra/flash'
+require 'jwt'
+
+CLIENT_ID = ENV['GH_BASIC_CLIENT_ID']
+CLIENT_SECRET = ENV['GH_BASIC_SECRET_ID']
+SESSION_SECRET = ENV['SESSION_SECRET']
+
+signing_key_path = File.expand_path('../app.rsa', __FILE__)
```
Comment (stahor) [inline app.rb]: put into config folder

File: app.rb [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+verify_key_path = File.expand_path('../app.rsa.pub', __FILE__)
+
+signing_key = ''
+verify_key = ''
+
+File.open(signing_key_path) do |file|
```
Comment (stahor) [inline app.rb]: why do you need path variable?

Comment (stahor) [inline app.rb]: why not just `OpenSSL::PKey.read(File.open('paht_to_file'))`?

File: app.rsa.example (hunk 8 above)
Comment (andersann) [inline app.rsa.example]: Yes, it's not real

File: app.rb (hunk 9 above)
Comment (andersann) [inline app.rb]: I wanted to be sure that it was being created

Comment (andersann) [inline app.rb]: I remove it

**Comment by andersann** [comment]
@stahor , added user's profile in new commit

File: app.rb (hunk 11 above)
Comment (stahor) [inline app.rb]: I don't actually understand why you need these keys.

Comment (andersann) [inline app.rb]: In [example](https://github.com/nickdufresne/jwt-sinatra-example/blob/master/app.rb) beside this lines is a comment 
```
######
# first we load up the private and public keys that we will use to sign and verify our JWT token
# using RS256 algo
######
```

Comment (stahor) [inline app.rb]: u could do this easier but okay. Another variant is here https://auth0.com/blog/ruby-authentication-secure-rack-apps-with-jwt/

File: app.rb [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+CLIENT_SECRET = ENV['GH_BASIC_SECRET_ID']
+SESSION_SECRET = ENV['SESSION_SECRET']
+
+signing_key_path = File.expand_path('../app.rsa', __FILE__)
+verify_key_path = File.expand_path('../app.rsa.pub', __FILE__)
+
+signing_key = OpenSSL::PKey.read(File.open(signing_key_path))
+verify_key = OpenSSL::PKey.read(File.open(verify_key_path))
+
+set :signing_key, signing_key
```
Comment (stahor) [inline app.rb]: set :signing_key, OpenSSL::PKey.read(File.open(signing_key_path))

File: app.rb [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+      exp = Time.at(exp.to_i)
+      return false if Time.now > exp
+      @user_id = payload['user_id']
+    rescue JWT::DecodeError
+      return false
+    end
+  end
+
+  def create_session
+    headers = { exp: Time.now.to_i + 200 }
```
Comment (stahor) [inline app.rb]: remove whitespaces

File: app.rb [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+    token = JWT.encode({ user_id: @user.id }, settings.signing_key, 'RS256', headers)
```
Comment (stahor) [inline app.rb]: remove `{}`

File: app.rb [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+  header = JWT.decode(token, settings.verify_key, true)
+  user_id = header[0]['user_id']
+  @user = User.find_by(id: user_id)
+  erb :update
+end
+post '/update' do
+  token = session['access_token']
+  header = JWT.decode(token, settings.verify_key, true)
+  user_id = header[0]['user_id']
+  @user = User.find_by(id: user_id)
```
Comment (stahor) [inline app.rb]: write method that will fetch user for you

**Comment by andersann** [comment]
@stahor made changes

File: app.rb [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+      create_session
+      redirect '/users'
+    else
+      flash[:error] = 'This user is already exist!'
+      redirect '/signup'
+    end
+  end
+
+  def exist?(email)
+    User.find_by(email: email)
```
Comment (stahor) [inline app.rb]: `User.exists?(email: email)` [docs](https://apidock.com/rails/ActiveRecord/Base/exists%3F/class)
and this method becomes useless. Use code above instead

File: app.rb [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+      save_user(@user)
+    end
+  end
+
+  def current_user
+    protected!
+    token = session['access_token']
+    header = JWT.decode(token, settings.verify_key, true)
+    user_id = header[0]['user_id']
+    @user = User.find_by(id: user_id)
```
Comment (stahor) [inline app.rb]: call it @current_user and memoize [guide](http://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/)

Comment (andersann) [inline app.rb]: @stahor,it should looks like something this?
```
def current_user_id
    token = session['access_token']
    header = JWT.decode(token, settings.verify_key, true)
    user_id = header[0]['user_id']
  end

  def current_user
    @current_user ||= User.find_by(id: current_user_id)
  end
```

**Comment by stahor** [comment]
:+1:

**Pull request merged**